### PR TITLE
feature: display status with a different shape depending on the status (#17728)

### PR DESCRIPTION
**Pull request by Jiehong**

### CODE REVIEW

**Comment by Jiehong** [comment]
Test on list-dependencies (note the black square for some services):

```
[nix-shell:~/Dev/systemd]$ ./build/systemctl list-dependencies audit
audit.service
● ├─system.slice
● └─sysinit.target
●   ├─dev-hugepages.mount
●   ├─dev-mqueue.mount
●   ├─firewall.service
●   ├─kmod-static-nodes.service
●   ├─sys-fs-fuse-connections.mount
■   ├─sys-kernel-config.mount
●   ├─sys-kernel-debug.mount
●   ├─systemd-ask-password-console.path
■   ├─systemd-journal-catalog-update.service
●   ├─systemd-journal-flush.service
●   ├─systemd-journald.service
●   ├─systemd-modules-load.service
●   ├─systemd-random-seed.service
●   ├─systemd-sysctl.service
●   ├─systemd-timesyncd.service
●   ├─systemd-tmpfiles-setup.service
●   ├─systemd-udev-trigger.service
●   ├─systemd-udevd.service
■   ├─systemd-update-done.service
●   ├─systemd-update-utmp.service
●   ├─cryptsetup.target
●   ├─local-fs.target
●   │ ├─-.mount
●   │ ├─boot.mount
●   │ ├─home.mount
■   │ ├─systemd-fsck-root.service
●   │ └─systemd-remount-fs.service
●   └─swap.target
```

File: src/basic/unit-def.h [hunk 1]
```diff
@@ -4,6 +4,7 @@
 #include <stdbool.h>
 
 #include "macro.h"
+#include "locale-util.h"
```
Comment (bluca) [inline src/basic/unit-def.h:6]: please sort alphabetically

**Comment by bluca** [comment]
I like this - just one minor nit. Waiting for others to have a look as well.

**Comment by vcaputo** [comment]
The square vs. circle strikes me as too subtle, but it seems like a good use of otherwise empty space conceptually.

I'd prefer they be more differentiated and clearly negative for failed/error, like X vs. circle.

**Comment by Jiehong** [comment]
> The square vs. circle strikes me as too subtle, but it seems like a good use of otherwise empty space conceptually.
> 
> I'd prefer they be more differentiated and clearly negative for failed/error, like X vs. ✓

@vcaputo : I kindda agree with that, but I didn't want to depart too much from the current display. I first thought I might as well replace ● by ⯈, but I decided not to. I'd be happy to hear other point of views!

**Comment by vcaputo** [comment]
My thinking is if we're going to give meaning to the bullets, may as well lean into it and make that meaning clear.  If this output is viewed as API requiring stability, it'll break equally either way and we can't change it at all.  Otherwise, do it well...

**Comment by keszybz** [comment]
>  If this output is viewed as API requiring stability, it'll break equally either way and we can't change it at all.

The formatting in `status` is not under any stability guarantees. It is explicitly designed for humans to look at. Guarantees for stability applies to `show` and other APIs which should be used for scripting. (Though even there, we'll change things if the old way turns out to be seriously broken.)

>  My thinking is if we're going to give meaning to the bullets, may as well lean into it and make that meaning clear. 

Agreed. I don't think the square works well to say "this one is broken". Maybe use "●" and "×" ? Or '○' and '●' ?

**Comment by Jiehong** [comment]
Here are the current ideas and some new ones:

Proposition 1

```
|  UnitActiveState  |  Glyph |
|-------------------|--------|
| UNIT_ACTIVE       |  ●     |
| UNIT_RELOADING    |  ●     |
| UNIT_INACTIVE     |  ■     |
| UNIT_FAILED       |  ■     |
| UNIT_ACTIVATING   |  ●     |
| UNIT_DEACTIVATING |  ●     |
| UNIT_MAINTENANCE  |  ■     |
```

Proposition 2

```
|  UnitActiveState  |  Glyph |
|-------------------|--------|
| UNIT_ACTIVE       |  ●     |
| UNIT_RELOADING    |  ●     |
| UNIT_INACTIVE     |  ×     |
| UNIT_FAILED       |  ×     |
| UNIT_ACTIVATING   |  ●     |
| UNIT_DEACTIVATING |  ●     |
| UNIT_MAINTENANCE  |  ×     |
```

Proposition 3

```
|  UnitActiveState  |  Glyph |
|-------------------|--------|
| UNIT_ACTIVE       |  ✓     |
| UNIT_RELOADING    |  ✓     |
| UNIT_INACTIVE     |  ×     |
| UNIT_FAILED       |  ×     |
| UNIT_ACTIVATING   |  ✓     |
| UNIT_DEACTIVATING |  ✓     |
| UNIT_MAINTENANCE  |  ×     |
```

Proposition 4

```
|  UnitActiveState  |  Glyph |
|-------------------|--------|
| UNIT_ACTIVE       |  ●     |
| UNIT_RELOADING    |  ●     |
| UNIT_INACTIVE     |  ○     |
| UNIT_FAILED       |  ○     |
| UNIT_ACTIVATING   |  ●     |
| UNIT_DEACTIVATING |  ●     |
| UNIT_MAINTENANCE  |  ○     |
```

Proposition 5

```
|  UnitActiveState  |  Glyph |
|-------------------|--------|
| UNIT_ACTIVE       |  Ⓐ     |
| UNIT_RELOADING    |  Ⓡ     |
| UNIT_INACTIVE     |  Ⓘ     |
| UNIT_FAILED       |  Ⓕ     |
| UNIT_ACTIVATING   |  Ⓐ     |
| UNIT_DEACTIVATING |  Ⓓ     |
| UNIT_MAINTENANCE  |  Ⓜ     |
```

Proposition 6

```
|  UnitActiveState  |  Glyph |
|-------------------|--------|
| UNIT_ACTIVE       |  ●     |
| UNIT_RELOADING    |  ↻     |
| UNIT_INACTIVE     |  ○     |
| UNIT_FAILED       |  ×     |
| UNIT_ACTIVATING   |  ↻     |
| UNIT_DEACTIVATING |  ↺     |
| UNIT_MAINTENANCE  |  ○     |
```

Which one do you think is the best? Feel free to mix/match to create a new one.

**Comment by keszybz** [comment]
```
|  UnitActiveState  |  Glyph |
|-------------------|--------|
| UNIT_ACTIVE       |  ●     |
| UNIT_RELOADING    |  ↻     |
| UNIT_INACTIVE     |  ○     |
| UNIT_FAILED       |  ×     |
| UNIT_ACTIVATING   |  ↻     |
| UNIT_DEACTIVATING |  ↺     |
| UNIT_MAINTENANCE  |  ○     |
```
I like the last one, except for the glyphs chosen for ACTIVAING/DEACTIVATING. ↻ for RELOADING is a nice touch. I think ACTIVATING could be shown with the same glyph as as ACTIVE (Actually, units are usually more "active" when activating — often running at 100% CPU — then when active, when they are idling. So from a user POV, both states are rather similar.) Similarly for DEACTIVATING: after all the unit is busy running through shutdown procedures, so ● would also be appropriate. Or maybe use a different glyph for those two, e.g. ◎?

Alt proposal:
```
|  UnitActiveState  |  Glyph |
|-------------------|--------|
| UNIT_ACTIVE       |  ●     |
| UNIT_RELOADING    |  ↻     |
| UNIT_INACTIVE     |  ○     |
| UNIT_FAILED       |  ×     |
| UNIT_ACTIVATING   |  ◎     |
| UNIT_DEACTIVATING |  ◎     |
| UNIT_MAINTENANCE  |  ◍     |
```

**Comment by Jiehong** [comment]
@keszybz I like that idea. Let's call it Proposition 7.

Proposition 8 (simplification of 7):

```
|  UnitActiveState  |  Glyph |
|-------------------|--------|
| UNIT_ACTIVE       |  ●     |
| UNIT_RELOADING    |  ↻     |
| UNIT_INACTIVE     |  ○     |
| UNIT_FAILED       |  ×     |
| UNIT_ACTIVATING   |  ●     |
| UNIT_DEACTIVATING |  ●     |
| UNIT_MAINTENANCE  |  ◍     |
```

**Comment by poettering** [comment]
hmm, i'd prefer a combo of prop 2 and 4. i.e.:

|  UnitActiveState  |  Glyph |
|-------------------|--------|
| UNIT_ACTIVE       |  ●     |
| UNIT_RELOADING    |  ●     |
| UNIT_INACTIVE     |  ○     |
| UNIT_FAILED       |   ×     |
| UNIT_ACTIVATING   |  ●     |
| UNIT_DEACTIVATING |  ●     |
| UNIT_MAINTENANCE  |  ○     |

i.e. i think this is a bit more subtle.

**Comment by vcaputo** [comment]
Prop 7 looks good to me except why not use ↑ and ↓ for activating and deactivating?

**Comment by poettering** [comment]
hmm, or well, i see now that prop 8 is very similar, besides the ↻ thing. which probably makes sense.

The maintainance thing i don't like though. first of all it looks terrible on my terminal, but i am not even sure what the thing is supposed to mean...

hmm, and we shouldn't forget that the console font people signalled they'd try to follow our unicode use to figure out which glyphs to include among the 512 they have. I think it's not a good point to make to convince them to introduce ◍ just because of the maintainance state.

hence maybe:


UnitActiveState | Glyph
-- | --
UNIT_ACTIVE | ●
UNIT_RELOADING | ↻
UNIT_INACTIVE | ○
UNIT_FAILED | ×
UNIT_ACTIVATING | ●
UNIT_DEACTIVATING | ●
UNIT_MAINTENANCE | ○

**Comment by poettering** [comment]
> Prop 7 looks good to me except why not use ↑ and ↓ for activating and deactivating?

let's not get lost in glyphs. In particular as those arrows kinda suggest to look where they point, but htere's nothing there. I'd really stay away from such "outside" pointing arrows hence.

**Comment by vcaputo** [comment]
> > Prop 7 looks good to me except why not use ↑ and ↓ for activating and deactivating?
> 
> let's not get lost in glyphs. In particular as those arrows kinda suggest to look where they point, but htere's nothing there. I'd really stay away from such "outside" pointing arrows hence.

I see your point, but it strikes me as quite misleading to present deactivating as active.  Maybe arrows aren't great as you've pointed out, but *something* should be used to describe "imminently inactive".

**Comment by poettering** [comment]
> I see your point, but it strikes me as quite misleading to present deactivating as active. Maybe arrows aren't great as you've pointed out, but _something_ should be used to describe "imminently inactive".

Well, I think it matters to make visible "something is running" as opposed to "nothing is running" as opposed to "everytjing's fucked", that's why I posted my earlier proposal. I don't think we should get lost in trying to communicate every single state via a separate glyph. If you want to know the details, look at the actual state.

I mean, if you really want to communicate everything you could ask the question: why not provide a glyph for each substate (i.e. per-unit-type state) too?

**Comment by poettering** [comment]
maybe we should use `<blink>` to communicate start-up/shutdown ;-)

**Comment by bluca** [comment]
> maybe we should use `<blink>` to communicate start-up/shutdown ;-)

Can we play the Windows 95 startup sound via unicode?

**Comment by Jiehong** [comment]
We go from everything is a round circle to better distinctions, so anything is an improvement.

I agree that not each state is worth identifying (some are just transient).

I'm fine with implementing @poettering 's proposal (although, I'm not sure ↻ is very common in most fonts in consoles).

**Comment by vcaputo** [comment]
> Well, I think it matters to make visible "something is running" as opposed to "nothing is running" as opposed to "everytjing's fucked", that's why I posted my earlier proposal. I don't think we should get lost in trying to communicate every single state via a separate glyph. If you want to know the details, look at the actual state.

Maybe one would go drill into the actual state if some things were shown as unexpectedly transient vs. the false sense of stability conferred by showing transients as running in this view.

Flattening transients as stable is tantamount to the trend of largely useless automobile gauges flattening all operating temps into the same needle position.  By the time the needle starts moving, a blown head gasket is already in the mail.

File: man/systemctl.xml [hunk 2]
```diff
@@ -232,8 +232,9 @@ Jan 12 10:46:45 example.com bluetoothd[8900]: gatt-time-server: Input/output err
 
             <para>The dot ("●") uses color on supported terminals to summarize the unit state at a glance. White
             indicates an <literal>inactive</literal> or <literal>deactivating</literal> state. Red indicates a
-            <literal>failed</literal> or <literal>error</literal> state and green indicates an
-            <literal>active</literal>, <literal>reloading</literal> or <literal>activating</literal> state.
+            <literal>failed</literal> or <literal>error</literal> state (and it becomes a square ("■") in this case).
```
Comment (keszybz) [inline man/systemctl.xml]: Outdated.

**Review by keszybz — CHANGES_REQUESTED** [review]

File: man/systemctl.xml (hunk 2 above)
Comment (Jiehong) [reply man/systemctl.xml]: updated.

File: src/basic/locale-util.c [hunk 3]
```diff
@@ -356,6 +356,9 @@ const char *special_glyph(SpecialGlyph code) {
                         [SPECIAL_GLYPH_TREE_SPACE]              = "  ",
                         [SPECIAL_GLYPH_TRIANGULAR_BULLET]       = ">",
                         [SPECIAL_GLYPH_BLACK_CIRCLE]            = "*",
+                        [SPECIAL_GLYPH_WHITE_CIRCLE]            = "*",
+                        [SPECIAL_GLYPH_MULTIPLICATION_SIGN]     = "x",
+                        [SPECIAL_GLYPH_CIRCLE_ARROW]            = "*",
```
Comment (poettering) [inline src/basic/locale-util.c:361]: test-locale-util could use an update too

**Comment by keszybz** [comment]
Looks great.

I tested this also on the tty: ↻ is not shown, but it is replaced by a square, which is good enough (rather similar to what we have before). That's good enough.

**Review by bluca — APPROVED** [review]
Looks very nice, good stuff!

**Pull request merged**